File: frontend/lib/models/meal_plan.dart
import 'recipe.dart';

class MealPlan {
  final int id;
  final int userId;
  final int recipeId;
  final String plannedDate;
  final String mealType;
  final bool isCompleted;
  final String? completedAt;
  final int? userRating;
  final String? userNotes;
  final String? createdAt;
  final String? updatedAt;
  final Recipe? recipe; // Included when fetched with recipe details

  MealPlan({
    required this.id,
    required this.userId,
    required this.recipeId,
    required this.plannedDate,
    required this.mealType,
    required this.isCompleted,
    this.completedAt,
    this.userRating,
    this.userNotes,
    this.createdAt,
    this.updatedAt,
    this.recipe,
  });

  factory MealPlan.fromJson(Map<String, dynamic> json) {
    return MealPlan(
      id: json['id'] as int? ?? 0,
      userId: json['user_id'] as int? ?? 0,
      recipeId: json['recipe_id'] as int? ?? 0,
      plannedDate: json['planned_date'] as String? ?? '',
      mealType: json['meal_type'] as String? ?? '',
      isCompleted: json['is_completed'] as bool? ?? false,
      completedAt: json['completed_at'] as String?,
      userRating: json['user_rating'] as int?,
      userNotes: json['user_notes'] as String?,
      createdAt: json['created_at'] as String?,
      updatedAt: json['updated_at'] as String?,
      recipe: json['recipe'] != null ? Recipe.fromJson(json['recipe'] as Map<String, dynamic>) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'recipe_id': recipeId,
      'planned_date': plannedDate,
      'meal_type': mealType,
      'is_completed': isCompleted,
      'completed_at': completedAt,
      'user_rating': userRating,
      'user_notes': userNotes,
      'created_at': createdAt,
      'updated_at': updatedAt,
      if (recipe != null) 'recipe': recipe!.toJson(),
    };
  }
}

File: backend/app/models/meal_plan.py
from datetime import datetime
from app import db


class MealPlan(db.Model):
    __tablename__ = 'meal_plans'

    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    recipe_id = db.Column(db.Integer, db.ForeignKey('recipes.id'), nullable=False)

    # Planning details
    planned_date = db.Column(db.Date, nullable=False, index=True)
    meal_type = db.Column(db.String(50))  # breakfast, lunch, dinner, snack

    # Status
    is_completed = db.Column(db.Boolean, default=False)
    completed_at = db.Column(db.DateTime)

    # User feedback
    user_rating = db.Column(db.Integer)  # 1-5 stars
    user_notes = db.Column(db.Text)

    # Timestamps
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    recipe = db.relationship('Recipe', backref='meal_plans')

    def to_dict(self):
        """Convert meal plan to dictionary"""
        return {
            'id': self.id,
            'user_id': self.user_id,
            'recipe': self.recipe.to_dict(include_ingredients=False) if self.recipe else None,
            'planned_date': self.planned_date.isoformat(),
            'meal_type': self.meal_type,
            'is_completed': self.is_completed,
            'completed_at': self.completed_at.isoformat() if self.completed_at else None,
            'user_rating': self.user_rating,
            'user_notes': self.user_notes,
            'created_at': self.created_at.isoformat()
        }

    def __repr__(self):
        return f'<MealPlan {self.planned_date} - {self.meal_type}>'
